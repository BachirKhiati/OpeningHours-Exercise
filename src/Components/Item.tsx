import React from 'react';
import { Text, View } from 'react-native';
import I18n from 'react-native-i18n';

import styles from '../Styles/ItemStyles';
import Colors from '../Styles/Colors';
import { NewDatesObjectType } from '../Utils/Conversion';

interface ItemsProps {
    item: NewDatesObjectType;
}

export default function Item({ item }: ItemsProps): JSX.Element {
    const { day, time, isToday, isClosed } = item;
    return (
        <View testID="list-item" style={styles.container}>
            <View style={styles.dateTextView}>
                <Text testID="today" style={styles.dateText}>
                    {I18n.t(day)}
                </Text>
                {isToday && (
                    <Text style={styles.isTodayText}>{I18n.t('today')}</Text>
                )}
            </View>
            <View style={styles.timeTextWrapper}>
                {isClosed || !Array.isArray(time) ? (
                    <Text
                        testID="closed-text"
                        style={[
                            styles.timeText,
                            { color: isClosed ? Colors.gray3 : Colors.black }
                        ]}
                    >
                        {I18n.t('closed')}
                    </Text>
                ) : (
                    Array.isArray(time) &&
                    time.map(({ open, close }, key) => (
                        <View style={styles.timeTextView} key={`Items${key}`}>
                            <Text testID="open.hour">
                                {open.hour} {open.period}
                            </Text>
                            <Text>{' - '}</Text>
                            <Text>
                                {close.hour} {close.period}
                            </Text>
                        </View>
                    ))
                )}
            </View>
        </View>
    );
}
