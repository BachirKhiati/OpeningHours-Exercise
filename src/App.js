import React, {useState, useEffect, useRef} from 'react';
import {
  Text,
  View,
  Dimensions,
  StyleSheet,
  TouchableHighlight,
  StatusBar,
  Image,
} from 'react-native';
import Animated, {interpolate} from 'react-native-reanimated';
import {
  bInterpolate,
  bInterpolateColor,
  useTransition,
} from 'react-native-redash';

import Icon from 'react-native-vector-icons/dist/SimpleLineIcons';
import I18n from 'react-native-i18n';


import Chevron from './Components/Chevron';
import Colors from './Styles/Colors';
import styles from './Styles/AppStyles';
import {Convert} from './Utils/Conversion';
import Item from './Components/Items';

const {width: WIDTH, height: HEIGHT} = Dimensions.get('window');
const MIN_WIDTH = WIDTH * 0.6;
const MAX_WIDTH = WIDTH * 0.8;
const TOP_INT = 2 / WIDTH;
const BTN_HEIGHT = 60;
const LIST_HEIGHT = WIDTH - BTN_HEIGHT;

const AnimatedIcon = Animated.createAnimatedComponent(Icon);

function App() {
  const [showHours, setShowHours] = useState(false);
  const [list, setList] = useState([]);

  const mainTransition = useTransition(showHours);

  const heightContainerInterpolate = bInterpolate(mainTransition, BTN_HEIGHT, WIDTH);
  const heightListInterPolate = bInterpolate(mainTransition, 0, LIST_HEIGHT);
  const heightListOpacity = bInterpolate(mainTransition, 0, 1);
  const widthInterpolate = bInterpolate(mainTransition, MIN_WIDTH, MAX_WIDTH);
  const topInterpolate = interpolate(mainTransition, {inputRange: [0, TOP_INT], outputRange: [1, 0]});
  const backgroundInterpolate = bInterpolateColor(mainTransition, Colors.green, Colors.white);
  const textColorInterpolate = bInterpolateColor(mainTransition, Colors.white, Colors.black);

  useEffect(() => {
    setList(Convert());
  }, []);

  function onPressShow() {
    setShowHours(!showHours);
  }

  return (
    <View style={styles.container}>
      <StatusBar backgroundColor={'transparent'} />
      <Animated.View style={{top: topInterpolate}}>
        <TouchableHighlight style={styles.btnWrapper} onPress={onPressShow}>
          <Animated.View
            style={[
              styles.animationContainer,
              {
                height: heightContainerInterpolate,
                width: widthInterpolate,
                backgroundColor: backgroundInterpolate,
              },
            ]}>
            <View style={styles.visibleView}>
              <AnimatedIcon
                name={'clock'}
                color={textColorInterpolate}
                size={20}
              />
              <Animated.Text
                style={[
                  styles.visibleViewText,
                  {
                    color: textColorInterpolate,
                  },
                ]}>
                {I18n.t('opening_hours')}
              </Animated.Text>
              <Chevron transition={mainTransition} />
            </View>
            <Animated.View
              style={[
                styles.minimizedView,
                {
                  height: heightListInterPolate,
                  opacity: heightListOpacity,
                },
              ]}>
              {list.map((item, key) => (
                <Item {...{item, key}} />
              ))}
            </Animated.View>
          </Animated.View>
        </TouchableHighlight>
      </Animated.View>
    </View>
  );
}
export default App;
