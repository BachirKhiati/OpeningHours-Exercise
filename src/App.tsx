import React, { useState, useEffect } from 'react';
import {
    View,
    Dimensions,
    TouchableOpacity,
    TouchableHighlight,
    StatusBar,
    Text,
    ActivityIndicator
} from 'react-native';
import Animated from 'react-native-reanimated';
import {
    bInterpolate,
    bInterpolateColor,
    useTransition
} from 'react-native-redash';
import Icon from 'react-native-vector-icons/SimpleLineIcons';
import I18n from 'react-native-i18n';

import Chevron from './Components/Chevron';
import { Convert, NewDatesObjectType } from './Utils/Conversion';
import Item from './Components/Items';
import styles from './Styles/AppStyles';
import Colors from './Styles/Colors';

const { width: WIDTH } = Dimensions.get('window');
// width when view minimized
const MIN_WIDTH = WIDTH * 0.6;
// width when view maximized
const MAX_WIDTH = WIDTH * 0.8;
// Top Interpolation value used when maximizing the view
const TOP_INT = 2 / WIDTH;

// initial btn/view  height- minimized
const BTN_HEIGHT = 60;

// List height when view is maximized
const LIST_HEIGHT = WIDTH - BTN_HEIGHT;

// loop  for switching language
interface LanguageTypes {
    [name: string]: string;
}
const LOOP_LANGUAGE: LanguageTypes = {
    fi: 'en',
    en: 'fi'
};

// to be able to transition the Icon(font) color between white to black.
const AnimatedIcon = Animated.createAnimatedComponent(Icon);

function App() {
    // state
    const [showHours, setShowHours] = useState(false);
    const [list, setList] = useState<Array<NewDatesObjectType>>([]);
    const [language, setLanguage] = useState('en');

    // main transition
    const mainTransition = useTransition(showHours);

    // interpolations all based on "showHours" value"
    // Visual Effects
    const heightContainerInterpolate = bInterpolate(
        mainTransition,
        BTN_HEIGHT,
        WIDTH
    );
    const heightListInterpolate = bInterpolate(mainTransition, 0, LIST_HEIGHT);
    const heightListOpacity = bInterpolate(mainTransition, 0, 1);
    const widthInterpolate = bInterpolate(mainTransition, MIN_WIDTH, MAX_WIDTH);
    const topInterpolate = Animated.interpolate(mainTransition, {
        inputRange: [0, TOP_INT],
        outputRange: [1, 0]
    });
    const backgroundInterpolate = bInterpolateColor(
        mainTransition,
        Colors.green,
        Colors.white
    );
    const textColorInterpolate = bInterpolateColor(
        mainTransition,
        Colors.white,
        Colors.black
    );

    // fetch list
    useEffect(() => {
        setList(Convert());
    }, []);

    function onShowPressed() {
        setShowHours(!showHours);
    }

    function onLanguagePressed() {
        I18n.locale = LOOP_LANGUAGE[language];
        setLanguage(LOOP_LANGUAGE[language]);
    }

    if (list.length === 0) {
        return (
            <View style={styles.container}>
                <ActivityIndicator size="large" color={Colors.green} />
            </View>
        );
    }

    return (
        <View style={styles.container}>
            <StatusBar backgroundColor="transparent" barStyle="dark-content" />
            <Animated.View style={{ top: topInterpolate }}>
                <TouchableHighlight
                    style={styles.btnWrapper}
                    onPress={onShowPressed}
                >
                    <Animated.View
                        style={[
                            styles.animationContainer,
                            {
                                height: heightContainerInterpolate,
                                width: widthInterpolate,
                                backgroundColor: backgroundInterpolate
                            }
                        ]}
                    >
                        <View style={styles.visibleView}>
                            <AnimatedIcon
                                name="clock"
                                color={textColorInterpolate}
                                size={20}
                            />
                            <Animated.Text
                                style={[
                                    styles.visibleViewText,
                                    {
                                        color: textColorInterpolate
                                    }
                                ]}
                            >
                                {I18n.t('opening_hours')}
                            </Animated.Text>
                            <Chevron transition={mainTransition} />
                        </View>
                        <Animated.View
                            style={[
                                styles.minimizedView,
                                {
                                    height: heightListInterpolate,
                                    opacity: heightListOpacity
                                }
                            ]}
                        >
                            {list.map((item, key) => (
                                <Item {...{ item, key }} />
                            ))}
                        </Animated.View>
                    </Animated.View>
                </TouchableHighlight>
            </Animated.View>
            <TouchableOpacity
                onPress={onLanguagePressed}
                style={styles.btnFloating}
            >
                <Text style={styles.textFloating}>
                    {LOOP_LANGUAGE[language]}
                </Text>
            </TouchableOpacity>
        </View>
    );
}

export default App;
