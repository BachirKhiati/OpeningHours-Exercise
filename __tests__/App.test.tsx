import { Text } from 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import { render, cleanup } from '@testing-library/react-native';
import { mount, shallow } from 'enzyme';

import App from '../src/App';
import Item from '../src/Components/Items';
import { json1Output, json6Output } from './json/testExpected';

jest.mock('react-native-vector-icons');
jest.mock('react-native-i18n');
jest.mock('react-native-redash');
jest.mock('react-native-gesture-handler');
jest.mock('react-native-reanimated', () =>
    require('react-native-reanimated/mock')
);
jest.useFakeTimers();

beforeEach(() => {
    cleanup();
    jest.clearAllMocks();
});

describe('Test setup check', () => {
    it('Render Simple Component', () => {
        const { queryByTestId } = render(
            <Text testID="text">Hello, world!</Text>
        );
        expect(queryByTestId('text')).not.toBeNull();
    });
});

describe('Rendering', () => {
    it('should match to snapshot', () => {
        const component = shallow(<App />);
        expect(component).toMatchSnapshot();
    });

    it('App renders correctly', () => {
        const tree = renderer.create(<App />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe('Button Render', () => {
    test('Initial Height', async () => {
        const { getByTestId } = render(<App />);
        const animatedView = getByTestId('animated-view').props.children[0];
        expect(animatedView.props.style.height).toEqual(60);
    });
});

describe('Render', () => {
    test('Initial state, - Loading indicator', async () => {
        const wrapper = shallow(<App />);
        expect(
            wrapper.findWhere(
                node => node.prop('testID') === 'ActivityIndicator'
            )
        ).toBeTruthy();
        expect(wrapper.contains(<Item item={json6Output[0]} />)).toBe(false);
    });

    test('Initial state, -render items', async () => {
        const { getByTestId, queryByTestId } = render(<App />);
        expect(getByTestId('card-view')).toBeTruthy();
        expect(queryByTestId('ActivityIndicator')).toBeNull();
    });

    test('Should render 7 items', async () => {
        const { getAllByTestId } = render(<App />);
        expect(getAllByTestId('list-item').length).toEqual(7);
    });
});

describe('Render list items ', () => {
    test('should show open time', async () => {
        const { getByTestId, queryByTestId } = render(
            <Item item={json1Output[0]} key={0} />
        );
        expect(getByTestId('open.hour')).toBeTruthy();
        expect(queryByTestId('closed-text')).toBeNull();
    });

    test('should show closed', async () => {
        const { getByTestId, queryByTestId } = render(
            <Item item={json6Output[0]} key={0} />
        );
        expect(getByTestId('closed-text')).toBeTruthy();
        expect(queryByTestId('open.hour')).toBeNull();
    });
});
