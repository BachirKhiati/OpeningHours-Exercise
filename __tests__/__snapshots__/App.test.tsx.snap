// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering App renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFF",
      "flex": 1,
      "justifyContent": "center",
    }
  }
  testID="ActivityIndicator"
>
  <ActivityIndicator
    animating={true}
    color="#5BCB02"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;

exports[`Rendering should match to snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFF",
      "flex": 1,
      "justifyContent": "center",
    }
  }
  testID="ActivityIndicator"
>
  <ActivityIndicator
    animating={true}
    color="#5BCB02"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;
